pd/src/info/specific.rs

```diff
@@ -429,6 +495,80 @@ impl SpecificQueryService for Info {
         }
     }
 
+    #[instrument(skip(self, request))]
+    async fn swap_executions(
+        &self,
+        request: tonic::Request<SwapExecutionsRequest>,
+    ) -> Result<tonic::Response<Self::SwapExecutionsStream>, Status> {
+        let state = self.storage.latest_snapshot();
+        state
+            .check_chain_id(&request.get_ref().chain_id)
+            .await
+            .map_err(|e| tonic::Status::unknown(format!("chain_id not OK: {e}")))?;
+        let request_inner = request.into_inner();
+        let start_height = request_inner.start_height;
+        let end_height = request_inner.end_height;
+        let trading_pair = request_inner.trading_pair;
+
+        // Convert to domain type ahead of time if necessary
+        let trading_pair: Option<TradingPair> = if let Some(trading_pair) = trading_pair {
+            Some(trading_pair.try_into().expect("invalid trading pair"))
+        } else {
+            None
+        };
+
+        use penumbra_app::dex::state_key;
+
+        let s = state.prefix(&state_key::swap_executions());
+        Ok(tonic::Response::new(
+            s.filter_map(move |i: anyhow::Result<(String, SwapExecution)>| {
+                async move {
+                    if i.is_err() {
+                        return Some(Err(tonic::Status::unavailable(format!(
+                            "error getting prefix value from storage: {}",
+                            i.err().unwrap()
+                        ))));
+                    }
+
+                    let (key, swap_execution) = i.unwrap();
+                    let parts = key.split('/').collect::<Vec<_>>();
+                    let height = parts[2].parse().expect("height is not a number");
+                    let asset_1: asset::Id =
+                        parts[3].parse().expect("asset id formatted improperly");
+                    let asset_2: asset::Id =
+                        parts[4].parse().expect("asset id formatted improperly");
```

Comment:
After #2606 we could pull these out of the `SwapExecution` directly rather than parsing keys